### PR TITLE
[WIP] CMF / Dynamic Routing

### PR DESCRIPTION
**Pull request by aschempp**
> Continuing from contao/core-bundle#1653

This is a work in progress how to implement the CMF Dynamic Router to the Contao code base. It currently does nothing more than routing, there are no new features like real controllers for pages etc.

Feel free to review and test it, but there's a lot still to be changed 😊

**TODO:**
- [ ] If `prepend_locale` is enabled and a page without locale is accessed, the visitor was redirected. E.g. https://demo.contao.org/news.html => https://demo.contao.org/en/news.html. 
- [x] If no pages are defined in `tl_page`, the `NoRootPageFoundException` should be rendered
- [x] Because the `PrettyErrorScreenHandler` only handles Contao routes (`_scope=frontend/backend`), it is now never triggered on actual 404s.
- [x] ~should we add a redirect if root is set to use HTTPS?~ see #262 
- [x] is `/` really a valid alternative alias for the first page instead of `index`?

### CODE REVIEW

**Comment by aschempp** [comment]
To "successfully" run the functional tests, the following adjustments to core are necessary:
 - Remove `exit` in `Controller::redirect()`
 - Replace the `routing.yml` in functional tests with the one from `Resources/config`

The tests are successful so far as the following things are *up for discussion*:
 1. `PageRoot::generate()` throws a 303 instead of 302 redirect. This code was probably not reached previously because the URL matching already redirected, but Sf routing now simply uses the correct controller.
 2. The alias `/.html` or `/en/.html` was previously redirected to the first page. I think that's wrong, it should generate a 404 (Sf routing does that).
 3. The "no root page found" state currently renders a 404 instead of a 500. And probably does not render the correct "information page", because Sf just does not find a matching route. Not sure how we should handle that, because we would need to have a controller for that case.

File: core-bundle/src/Framework/ContaoFramework.php [hunk 1]
```diff
@@ -80,9 +75,8 @@ class ContaoFramework implements ContaoFrameworkInterface, ContainerAwareInterfa
      */
     private $hookListeners = [];
 
-    public function __construct(RequestStack $requestStack, RouterInterface $router, ScopeMatcher $scopeMatcher, string $rootDir, int $errorLevel)
+    public function __construct(?RequestStack $requestStack /* removed in Contao 4.7 */, RouterInterface $router, ScopeMatcher $scopeMatcher, string $rootDir, int $errorLevel)
```
Comment (leofeyer) [inline core-bundle/src/Framework/ContaoFramework.php]: Why did you comment this instead of removing it?

File: core-bundle/src/Framework/ContaoFramework.php [hunk 2]
```diff
@@ -127,6 +118,25 @@ public function setHookListeners(array $hookListeners): void
         $this->hookListeners = $hookListeners;
     }
 
+    public function setRequest(Request $request)
```
Comment (leofeyer) [inline core-bundle/src/Framework/ContaoFramework.php]: Why is this necessary?

File: core-bundle/src/Framework/ContaoFramework.php [hunk 3]
```diff
@@ -158,26 +168,36 @@ public function getAdapter($class): Adapter
...
         }
 
-        \define('TL_START', microtime(true));
-        \define('TL_ROOT', $this->rootDir);
-        \define('TL_REFERER_ID', $this->getRefererId());
+        if (!\defined('TL_START')) {
+            \define('TL_START', microtime(true));
+        }
+
+        if (!\defined('TL_ROOT')) {
```
Comment (leofeyer) [inline core-bundle/src/Framework/ContaoFramework.php]: This check should not have to be there, as the framework cannot be initialized twice. Did you add it for the unit tests only?

File: core-bundle/src/Routing/FrontendLoader.php [hunk 4]
```diff
@@ -61,7 +67,7 @@ public function supports($resource, $type = null): bool
      */
     private function addFrontendRoute(RouteCollection $routes, array $defaults): void
     {
-        $route = new Route('/{alias}%contao.url_suffix%', $defaults, ['alias' => '.+']);
+        $route = new Route('/{alias}'.$this->urlSuffix, $defaults, ['alias' => '.+']);
```
Comment (leofeyer) [inline core-bundle/src/Routing/FrontendLoader.php:70]: What was the problem with the previous implementation?

File: core-bundle/src/Routing/LegacyRouteProvider.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        return $this->routeProvider->getRouteCollectionForRequest($request);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteByName($name)
+    {
+        if ('contao_frontend' === $name || 'contao_index' === $name) {
```
Comment (leofeyer) [inline core-bundle/src/Routing/LegacyRouteProvider.php:51]: Why can we no longer have these routes in the `routing.yml` file?

**Review by leofeyer — CHANGES_REQUESTED** [review]
It seems to me that you have made some changes for the unit tests only, which is the wrong approach IMHO.

File: core-bundle/src/Framework/ContaoFramework.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/Framework/ContaoFramework.php]: Because it would be a BC break?

File: core-bundle/src/Framework/ContaoFramework.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Framework/ContaoFramework.php]: If the framework is bootest before routing (a listener with a higher priority) we cannot initialize some parts of the framework.

File: core-bundle/src/Framework/ContaoFramework.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/src/Framework/ContaoFramework.php]: `TL_ROOT` cannot be initialized if the request is not matched.

File: core-bundle/src/Routing/FrontendLoader.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/Routing/FrontendLoader.php:70]: Because we're no longer using the route loader but manually use the class (to keep BC) the container parameters are not resolved. see https://github.com/contao/contao/pull/95/files#diff-5f8c741675801f711d0a61e756575413R51

File: core-bundle/src/Routing/LegacyRouteProvider.php (hunk 5 above)
Comment (aschempp) [reply core-bundle/src/Routing/LegacyRouteProvider.php:51]: They must not be matched by the routing, but they must be available to the URL generator.

File: core-bundle/src/Framework/ContaoFramework.php (hunk 1 above)
Comment (leofeyer) [reply core-bundle/src/Framework/ContaoFramework.php]: Hm, the parameter has been removed in Contao 4.7 and your pull request is targeted against Contao 4.7. 🤔

Comment (aschempp) [reply core-bundle/src/Framework/ContaoFramework.php]: What do you mean it has been removed?

Comment (leofeyer) [reply core-bundle/src/Framework/ContaoFramework.php]: That's what your comment says: `/* removed in Contao 4.7 */`

Comment (aschempp) [reply core-bundle/src/Framework/ContaoFramework.php]: lol. Yeah, it was removed **in this pull request** because we now have `setRequest`. But we can't change the arguments, that's why there is a comment on why the argument is still there.

**Comment by leofeyer** [comment]
> If `prepend_locale` is enabled and a page without locale is accessed, the visitor was redirected. E.g. https://demo.contao.org/news.html => https://demo.contao.org/en/news.html. Is that intended?

It sure is. I have added a test in https://github.com/contao/contao/commit/825a5b095b84415b88645a38dc3bbcfe8e5a9d85.

**Comment by ausi** [comment]
This redirect is currently also triggered for non-existent pages (e.g. <https://demo.contao.org/foo.html> redirects to <https://demo.contao.org/en/foo.html>). IMO it would be better to show a 404 page directly instead of the redirect.

**Comment by leofeyer** [comment]
> To "successfully" run the functional tests, the following adjustments to core are necessary:
> - Remove `exit` in `Controller::redirect()`
> - Replace the `routing.yml` in functional tests with the one from `Resources/config`

Done.

> The tests are successful so far as the following things are *up for discussion*:
> 1. `PageRoot::generate()` throws a 303 instead of 302 redirect. This code was probably not reached previously because the URL matching already redirected, but Sf routing now simply uses the correct controller.

Fine with me.

>  2. The alias `/.html` or `/en/.html` was previously redirected to the first page. I think that's wrong, it should generate a 404 (Sf routing does that).

Perfectly fine!

>  3. The "no root page found" state currently renders a 404 instead of a 500. And probably does not render the correct "information page", because Sf just does not find a matching route. Not sure how we should handle that, because we would need to have a controller for that case.

As discussed in Mumble on October 18th, it would be good to preserve the "no root page found" error if possible, so beginners know why they don't see anything.

**Comment by aschempp** [comment]
PR updated to correctly sort/match pages based on the browser accept language. That was not present yet 😇

**Comment by aschempp** [comment]
> As discussed in Mumble on October 18th, it would be good to preserve the "no root page found" error if possible, so beginners know why they don't see anything.

This is possible, however, the `PrettyErrorScreenListener` does not render at all if the pages are not in `scope = frontend/backend` (https://github.com/contao/contao/blob/master/core-bundle/src/EventListener/PrettyErrorScreenListener.php#L106). And if the router does not match a page, there is no scope, so it never renders. Can we agree on changing this so the pretty screens are always rendered for HTML output? @sheeep @bytehead

**Comment by leofeyer** [comment]
I'm not sure about it, but if it is the only solution, I'd rather have it this way than not having it at all.

**Comment by bytehead** [comment]
The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them. 

> Can we agree on changing this so the pretty screens are always rendered for HTML output?

👎

**Comment by aschempp** [comment]
> The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them.

How do you want to know if a 404 (no match found) is Contao route?

**Comment by aschempp** [comment]
> The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them.

How do you want to know if a 404 (no match found) is Contao route?

**Comment by aschempp** [comment]
> The `PrettyErrorScreenListener` is fine for me for routes in Contao scope, but only for them.

How do you want to know if a 404 (no match found) is Contao route?

**Comment by leofeyer** [comment]
Yeah, that seems impossible. You can either enable pretty error screens – then you will get them everywhere – or you can disable them – then you will not get them anywhere.

**Comment by leofeyer** [comment]
@aschempp Regarding your new questions:

> should we add a redirect if root is set to use HTTPS?

Yes, that would be nice. 👍 

> is / really a valid alternative alias for the first page instead of index?

I guess so, as some servers distinct between `/home` and `/home/` (aka with or without trailing slash).

**Comment by aschempp** [comment]
> should we add a redirect if root is set to use HTTPS?

I think that should be a new feature that we implement separately from this PR

> > is / really a valid alternative alias for the first page instead of index?

> I guess so, as some servers distinct between /home and /home/ (aka with or without trailing slash).

I don't think you got this right. The alias "/" (no other content) is matched as the index page in the Contao 3 router.

**Comment by leofeyer** [comment]
Yes, I know. This is intentionally.

**Comment by aschempp** [comment]
I guess this is *up for discussion* again as we need to agree about the `PrettyErrorScreenListener`

**Comment by bytehead** [comment]
> Yeah, that seems impossible. You can either enable pretty error screens – then you will get them everywhere – or you can disable them – then you will not get them anywhere.

If the `PrettyErrorScreenListener` is disabled by default, I'm fine with this then.

**Comment by aschempp** [comment]
> If the PrettyErrorScreenListener is disabled by default, I'm fine with this then.

How can it be disabled by default? A Contao 404 page would never be rendered, which is a basic feature of every Contao website.

**Comment by bytehead** [comment]
> How can it be disabled by default?

It has to be disabled while using Contao just as a Symfony bundle.
I don't care what's configured by default in the _Managed Edition_.

> A Contao 404 page would never be rendered, which is a basic feature of every Contao website.

What if a 404 page (type) is configured in the page tree?

**Comment by aschempp** [comment]
> What if a 404 page (type) is configured in the page tree?

Thats what I'm talking about. It would never be rendered …

**Comment by bytehead** [comment]
Why not? If you have a configured 404 page in the page tree, you don't need to have the `PrettyErrorScreenListener`, don't you?

**Comment by aschempp** [comment]
Yes you do. The `PrettyErrorScreenHandler` catches `ResourceNotFoundException` or `HttpNotFoundException` and renders the page from page tree if available 😉

**Comment by bytehead** [comment]
Ok. Fine for me in the _Managed Edition_. I don't like to leak Contao too much into Symfony as just a bundle (as we use it in all Contao cases).

If I'm fine with the Contao 404, I'll have to configure `prettyErrorScreen: true`, by default I get the default Symfony one.

**Comment by leofeyer** [comment]
As discussed in Mumble on November, 22rd:

> [15:44] (Channel) andy: es ist ganz einfach
> [15:44] (Channel) andy: core-bundle default aus
> [15:44] (Channel) andy: manager-bundle schaltet es ein

**Comment by leofeyer** [comment]
@aschempp Can you please fix the existing unit tests and add the missing ones (see coverage)? Then I can merge this PR.

File: core-bundle/src/Resources/config/listener.yml [hunk 6]
```diff
@@ -70,10 +70,21 @@ services:
 
     contao.listener.exception_converter:
         class: Contao\CoreBundle\EventListener\ExceptionConverterListener
+        arguments:
+            - "@database_connection"
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/listener.yml:74]: I assume you have check that we are not ending in a loop if the database class throws an exception while we are converting the exception?

File: core-bundle/src/Resources/contao/classes/Frontend.php [hunk 7]
```diff
@@ -59,6 +59,8 @@ public function __construct()
 	 */
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/Frontend.php:65]: The phpDoc is missing a `@deprecated` tag (this is just a note to myself).

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+        if (!empty($defaults['_locale']) && $this->configAdapter->get('addLanguageToUrl')) {
+            $this->inputAdapter->setGet('language', $defaults['_locale']);
+        }
+
+        if (empty($defaults['parameters'])) {
+            return $defaults;
+        }
+
+        // TODO should we use trim() instead of substr()?
```
Comment (leofeyer) [inline core-bundle/src/Routing/Enhancer/InputEnhancer.php]: Yes, `ltrim()` I guess.

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        $fragments = explode('/', substr($defaults['parameters'], 1));
+
+        // Add the second fragment as auto_item if the number of fragments is even
+        if ($this->configAdapter->get('useAutoItem') && 0 !== \count($fragments) % 2) {
+            array_unshift($fragments, 'auto_item');
+        }
+
+        for ($i = 0, $c = \count($fragments); $i < $c; $i += 2) {
+            // Skip key value pairs if the key is empty (see #4702)
+            // TODO why does this make sense?
```
Comment (leofeyer) [inline core-bundle/src/Routing/Enhancer/InputEnhancer.php]: I guess that is the double slash case `news//something.html`. I will check if the functional tests handle this case.

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            if ('' === $fragments[$i]) {
+                continue;
+            }
+
+            // Abort if there is a duplicate parameter (duplicate content) (see #4277)
+            // TODO should we use $request->query here?
```
Comment (leofeyer) [inline core-bundle/src/Routing/Enhancer/InputEnhancer.php]: Yes

File: core-bundle/src/Routing/Matcher/LegacyMatcher.php [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            array_splice($fragments, 1, 0, ['auto_item']);
+        }
+
+        return $fragments;
+    }
+
+    private function executeLegacyHook(array $fragments)
+    {
+        foreach ($GLOBALS['TL_HOOKS']['getPageIdFromUrl'] as $callback) {
+            /** @noinspection StaticInvocationViaThisInspection */
```
Comment (leofeyer) [inline core-bundle/src/Routing/Matcher/LegacyMatcher.php]: I think we have agreed on not adding any IDE specific code. 🙈

File: core-bundle/src/Routing/RouteProvider.php [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+        $table = $this->pageAdapter->getTable();
+
+        foreach ($names as $name) {
+            if (0 !== strncmp($name, $table.'.', 8)) {
+                continue;
+            }
+
+            [, $id] = explode('.', $name);
+
+            $ids[] = $id;
```
Comment (leofeyer) [inline core-bundle/src/Routing/RouteProvider.php]: We should type-cast to `(int)` here.

File: core-bundle/src/Routing/RouteProvider.php [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+            $requirements['_locale'] = $page->rootLanguage;
+        }
+
+        $routes[$table.'.'.$page->id] = new Route(
+            $path,
+            $defaults,
+            $requirements,
+            ['utf8' => true],
+            $page->domain,
+            null // TODO should we match SSL only if enabled in root? => $page->rootUseSSL ? 'https' : null
```
Comment (leofeyer) [inline core-bundle/src/Routing/RouteProvider.php]: I would prefer to automatically redirect HTTP to HTTPS.

File: core-bundle/tests/Framework/ContaoFrameworkTest.php [hunk 14]
```diff
@@ -277,6 +260,10 @@ public function testOverridesTheErrorLevel(): void
         error_reporting($errorReporting);
     }
 
+    /**
+     * @runInSeparateProcess
```
Comment (leofeyer) [inline core-bundle/tests/Framework/ContaoFrameworkTest.php:310]: Why did you re-add this? The tests in the master branch work without a separate process.

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php (hunk 10 above)
Comment (aschempp) [reply core-bundle/src/Routing/Enhancer/InputEnhancer.php]: See 847914e2ed2909575f53e21e66a8e9f42ed40004

File: core-bundle/src/Routing/Matcher/LegacyMatcher.php (hunk 11 above)
Comment (aschempp) [reply core-bundle/src/Routing/Matcher/LegacyMatcher.php]: Removed in 847914e2ed2909575f53e21e66a8e9f42ed40004

File: core-bundle/src/Routing/RouteProvider.php (hunk 12 above)
Comment (aschempp) [reply core-bundle/src/Routing/RouteProvider.php]: Implemented in 847914e2ed2909575f53e21e66a8e9f42ed40004

File: core-bundle/src/Routing/RouteProvider.php (hunk 13 above)
Comment (aschempp) [reply core-bundle/src/Routing/RouteProvider.php]: see #262

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php (hunk 9 above)
Comment (aschempp) [reply core-bundle/src/Routing/Enhancer/InputEnhancer.php]: see #263

File: core-bundle/src/Routing/Enhancer/InputEnhancer.php (hunk 8 above)
Comment (aschempp) [reply core-bundle/src/Routing/Enhancer/InputEnhancer.php]: As discussed, this should not be changed to not remove multiple slashes on the beginning of the path.

File: core-bundle/src/Resources/contao/classes/Frontend.php (hunk 7 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Frontend.php:65]: Added in 847914e2ed2909575f53e21e66a8e9f42ed40004

File: core-bundle/src/Resources/config/listener.yml (hunk 6 above)
Comment (aschempp) [reply core-bundle/src/Resources/config/listener.yml:74]: I have not, but I'm pretty sure the Symfony exception handler handles this case. But feel free to verify.

**Comment by asaage** [comment]
Can this accomplish correct error-pages for sites with enabled folder-url's?
https://github.com/contao/core-bundle/issues/1624

File: core-bundle/tests/Functional/RoutingTest.php [hunk 15]
```diff
@@ -189,8 +189,8 @@ public function getAliases(): array
             ],
             'Renders the default page if the alias is empty' => [
```
Comment (leofeyer) [inline core-bundle/tests/Functional/RoutingTest.php]: Renders the 404 page if the alias is empty

File: core-bundle/tests/Functional/RoutingTest.php [hunk 16]
```diff
@@ -412,9 +412,9 @@ public function getAliasesWithLocale(): array
...
```
Comment (leofeyer) [inline core-bundle/tests/Functional/RoutingTest.php]: Renders the 404 page if the alias is empty

File: core-bundle/tests/Functional/RoutingTest.php [hunk 17]
```diff
@@ -770,8 +770,8 @@ public function getRootAliases(): array
             ],
             'Throws a "no root page found" exception if no language matches' => [
```
Comment (leofeyer) [inline core-bundle/tests/Functional/RoutingTest.php]: Renders the 404 page…

File: core-bundle/tests/Functional/RoutingTest.php (hunk 15 above)
Comment (aschempp) [reply core-bundle/tests/Functional/RoutingTest.php]: Fixed in 5496681b13cec9843806184597d48623796eab1c

File: core-bundle/tests/Functional/RoutingTest.php (hunk 16 above)
Comment (aschempp) [reply core-bundle/tests/Functional/RoutingTest.php]: Fixed in 5496681b13cec9843806184597d48623796eab1c

File: core-bundle/tests/Functional/RoutingTest.php (hunk 17 above)
Comment (aschempp) [reply core-bundle/tests/Functional/RoutingTest.php]: Fixed in 5496681b13cec9843806184597d48623796eab1c

File: core-bundle/src/Resources/contao/classes/Frontend.php [hunk 18]
```diff
@@ -291,84 +299,71 @@ public static function getRootIdFromUrl()
...
 	public static function getRootPageFromUrl()
 	{
-		// HOOK: add custom logic
-		if (isset($GLOBALS['TL_HOOKS']['getRootPageFromUrl']) && \is_array($GLOBALS['TL_HOOKS']['getRootPageFromUrl']))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/classes/Frontend.php:295]: Removing the hook seems like a BC break to me.

Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Frontend.php:295]: The hook is triggered by the the new route provider (https://github.com/contao/contao/pull/95/files#diff-8f3e1abe0064e8534f4e9193ef359276R470)

**Comment by leofeyer** [comment]
Thanks a lot @aschempp. This is huge! 🎉
